Config setting improvements and `CONFIG` command fixes

johnnovak opened the pull request:
# Description

## Fixes

- `CONFIG -h viewport` displayed no description because the `%` signs inside the help description confused `WriteOut()` which has printf-like behaviour. The change in behaviour was most likely unintentionally introduced by @japsmits config-related improvements (to be fair, this would be hard to spot as it only affects setting descriptions that include `%` signs in a very specific way).
- Therefore, `%` signs are now encoded as `%%` in all config descriptions, and the config writer handles them appropriately.
- Fixed `CONFIG -h SEC PROP` resulting in a crash if _PROP_ is non-existent (regression introduced in eee8e2c8050d4)
- `CONFIG -h SEC PROP` printed two error messages if both _SEC_ and _PROP_ are non-existent—now it only prints a single error.

## Cosmetic changes

I've shuffled the order of a few config settings around; the relevant commit messages should make everything self-explanatory.

I _almost_ reordered some sections too, but that's risky as the init functions are called in registration order. So changing the section order is just the ticket to potentially introduce subtle bugs here and there, or outright crashes when changing certain settings at runtime.

Reordering properties within sections, on the other hand, is safe.

I've also added newlines before and after the output of the `CONFIG -h setting_name` command:

<img width="1215" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/02b602da-ba83-4396-b56c-36112bbf40d0">

# Manual testing

Started up Staging after each change, then ran a few games at the the end.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

weirddan455 reviewed (approved):
Looks like a solid improvement :+1:

FeralChild64 commented on src/dosbox.cpp:
```diff
@@ -1331,8 +1331,8 @@ void DOSBOX_Init()
 	MSG_Add("AUTOEXEC_CONFIGFILE_HELP",
-	        "Lines in this section will be run at startup.\n"
-	        "You can put your MOUNT lines here.\n");
+	        "Every line in this section is executed at startup as DOS commands.\n"
```
Maybe:

_Every line in this section is executed at startup as a DOS command._

or

_All the lines in this section are executed at startup as DOS commands._

johnnovak replied:
Nice catch, I'll take the first 😄

FeralChild64 commented on src/gui/sdlmain.cpp:
```diff
@@ -4213,14 +4213,65 @@ static void config_add_sdl()
+	pstring->SetOptionHelp("opengl_default",
+	                       "Rendering backend to use for graphics output ('opengl' by default).\n"
+	                       "Only the 'opengl' backend has shader support and is thus the preferred option.\n"
+						   "The 'texture' backend is only provided as a last resort fallback for buggy or\n"
```
Indentation is wrong here.

johnnovak commented:
@FeralChild64 @weirddan455 I've discovered some regressions when addressing the PR comments, so I've added a couple of more commits and I've extended the top description. Ready for another round of review, gents 😎

weirddan455 commented on src/gui/sdlmain.cpp:
```diff
@@ -4306,10 +4306,10 @@ static void config_add_sdl()
 	        "From 0 (no transparency) to 90 (high transparency).");
 
-	pstring = sdl_sec->Add_path("max_resolution", deprecated, "");
+	pstring = sdl_sec->Add_string("max_resolution", deprecated, "");
```
This and `viewport_resolution` appear to be unused.  I would expect `deprecated` to do something.  Otherwise why not just remove it?  Are we missing a call to `SetDeprecatedWithAlternateValue` here?

The replacement config value `viewport` also has this incorrect path type by the way:

https://github.com/dosbox-staging/dosbox-staging/blob/3157a8c7f7604430e7e97b8715cbb22a0c84ff9f/src/gui/render.cpp#L1064

johnnovak replied:
Try `config -h max_resolution`, or put `max_resolution` in your config, and you'll see why they need to stay.

We want to inform people about deprecated settings in their configs and give helpful tips on how to migrate to the new settings.

johnnovak replied:
Also changing the type of `viewport` to string, good pickup 🎉

FeralChild64 commented on src/gui/render.cpp:
```diff
@@ -1108,12 +1108,12 @@ static void init_render_settings(Section_prop& secprop)
+	        "  N%%:              Similar to 'WxH' but the size is specified as a percentage\n"
 	        "                   of the desktop size.\n"
-	        "  relative H% V%:  The viewport is set to a 4:3 aspect ratio rectangle fit into\n"
+	        "  relative H%% V%%:  The viewport is set to a 4:3 aspect ratio rectangle fit into\n"
```
Will this work with `blocksize`, `prebuffer`, and similar values, having `%s` in the config option help string?

weirddan455 replied:
Oh, maybe it's more complicated than I thought.  `GetHelp()` also parses the % signs.

https://github.com/dosbox-staging/dosbox-staging/blob/3157a8c7f7604430e7e97b8715cbb22a0c84ff9f/src/misc/setup.cpp#L304-L315

My suggested change is still better than what we have.  It avoids parsing the string a 2nd time for no reason.  But it might not solve the problem if it's happening inside `GetHelp()`.

If that's the case then yeah I guess we need the double %% sign.

johnnovak replied:
Yeah we need the `%%` signs. I've added some explanatory comments to `GetHelp()`, plus I've ditched `replace_all` in favour of `format_str`. Keeping `replace_all`, though, as it's a useful string util function to have.

johnnovak commented:
Okay @weirddan455 & @FeralChild64, I've addressed the comments again and yet again made some further minor improvements in additional commits if you wanna take another look.

weirddan455 reviewed (approved):
Looks good

FeralChild64 reviewed (approved):
Looks OK, I have no more remarks.

johnnovak commented:
Thanks @FeralChild64 & @weirddan455 for the review. Another big one down 🎉
